enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,108 @@
+are making progress or if their state causes the cluster to be not upgradeable. These conditions are essential for
+cluster administrators to determine the health of their clusters, and they are usually the first thing Red Hat support
+is going to look at in case there is a problem with the cluster.
+
+Unfortunately, in some edge cases, these conditions might not reflect the current state of the cluster. For example,
+since they depend on the running operator, these conditions might become stale and outdated if the operator is not
+running or cannot contact the API server. And while the API server might still work, the result
+of `oc get clusteroperators` does not represent the cluster's current state.
+
+Therefore, we need a detection mechanism to help detect in-active operators or operators that are not actively updating
+the conditions.
+
+## Motivation
+
+In the past, there were multiple cases when either kubelet or CRI-O failed to run pods, either by KCM malfunctioning or
+some other system error. However, the static pods succeeded in running, making the Kubernetes API server available and
+accessible with `oc`. When support or engineers ran `oc get clusteroperators` against this cluster, they saw all
+operators reporting available, non-degraded and upgradeable status, but the truth was that none of the operators that
+provided these conditions were running.
+
+### Goals
+
+* Detect inactive cluster operators
+* Provide more accurate condition messages and status
+
+### Non-Goals
+
+* Make a new operator that detects other operators status
+* Create a heart-beat mechanism
+
+## Proposal
+
+To implement stale operator detection, we need a controller process that will run close to the Kubernetes API server,
+preferably using a local connection and client-based certificate for authentication to avoid dependencies on working
+auth and network when the cluster is unstable.
+
+The controller will watch the clusteroperator resource and check the `status.[]conditions.lastTransitionTime` field for
+Degraded, Available, Progressing, and Upgradeable conditions. Suppose the "lastTransitionTime" field is not updated for
+10 minutes. In that case, the controller will change the human-readable condition message
+to: `Operator checking for stale status, the active operator will reset this message: <original message>`.
+
+After the message is changed, as it suggests, the active operators should reset this message back to the original
+message, removing the "checking" part. This shows a healthy, active operator.
+
+Suppose the operator is not active and does not reset the message back to the original after 10 minutes. In that case,
+the controller will change the message again
+to: `Operator has not updated this condition for more than 20 minutes, last known condition state was "<False/True>", original message: <original message>`
+. In addition to changing the message, this time, the controller will also change the condition status from False or
```

Comment:
if we decide to take this controller to production, the flapping messages are an inconvenience to the cluster operators.
I know we can't add new fields to `Condition`, but we could introduce new fields directly under `Status`, like `lastUpdatedTime` which the operator will need to update every `10m` or something - this has the same runtime overhead as flapping the message.

Only trade off is each operator needs to incorporate this heart-beat pattern into its logic, which we could make less of a maintenance burden if every operator in question uses a shared `co` status updater. 
I think it's reasonable to accept this trade-off of additional maintenance burden on the operator in order to give cluster admins a better experience

Reply:
> Only trade off is each operator needs to incorporate this heart-beat pattern into its logic, which we could make less of a maintenance burden if every operator in question uses a shared co status updater.
I think it's reasonable to accept this trade-off of additional maintenance burden on the operator in order to give cluster admins a better experience

This is acceptable to me, but will take a significant effort to push.  Since the apiserver team often fields these problems due to every operator reporting available, I'll leave the choice to you
1. flap the message in a non-destructive way and address the problem immediately
2. take ownership of the feature, add a new field, update every operator manage that new field, then write a controller to use this field.

I have no preference, but I do offer a complete solution, mergeable today for option 1.

Reply:
I'll throw out another possibility if someone wanted to build a cleaner thing: operators should have unexpired leases.  Given a relatedResource in status that is a lease, a clever controller could go from clusteroperator->relatedresource->lease->expired or not.  that would be peer to the current heatbeat mechanism for kubelets for instance.

I'm willing to back any of the three ideas listed in this thread so far.  But I think option 2 will have to be led by someone other than me.  option three might be an interesting compromise.